src/app/settings/admin/presets/dialog/steps/settings/provider/openstack/component.ts

```diff
@@ -87,17 +101,55 @@ export class OpenstackSettingsComponent extends BaseFormValidator implements OnI
-      username: this.form.get(Controls.Username).value,
-      password: this.form.get(Controls.Password).value,
-      project: this.form.get(Controls.Project).value,
-      projectID: this.form.get(Controls.ProjectID).value,
+    const preset = {
       domain: this.form.get(Controls.Domain).value,
       network: this.form.get(Controls.Network).value,
       securityGroups: this.form.get(Controls.SecurityGroups).value,
       floatingIPPool: this.form.get(Controls.FloatingIPPool).value,
       routerID: this.form.get(Controls.RouterID).value,
       subnetID: this.form.get(Controls.SubnetID).value,
     } as OpenstackPresetSpec;
+
+    // User can only specify credentials for a single credential type i.e. either default or application
+    if (this.credentialsType === CredentialsType.Default) {
+      preset.username = this.form.get(Controls.Username).value;
+      preset.password = this.form.get(Controls.Password).value;
+      preset.project = this.form.get(Controls.Project).value;
+      preset.projectID = this.form.get(Controls.ProjectID).value;
+    } else if (this.credentialsType === CredentialsType.Application) {
+      preset.applicationCredentialID = this.form.get(Controls.ApplicationCredentialID).value;
+      preset.applicationCredentialSecret = this.form.get(Controls.ApplicationCredentialSecret).value;
+    }
+    this._presetDialogService.preset.spec.openstack = preset;
+  }
+
+  changeView(event: MatButtonToggleChange): void {
+    this.credentialsType = event.value;
+    this.updateValidators();
+  }
+
+  updateValidators(): void {
+    // Conditional rendering using ngIf only effects the rendered content; html. To make sure that the validation for our
+    // form is updated when changing view we have to update validators from code-behind.
+    const validatorDefaultCredentials = this.credentialsType === CredentialsType.Default ? [Validators.required] : null;
+    const validatorAppCredentials = this.credentialsType === CredentialsType.Application ? [Validators.required] : null;
+
+    this.form.get(Controls.Username).setValidators(validatorDefaultCredentials);
+    this.form.get(Controls.Username).updateValueAndValidity();
+
+    this.form.get(Controls.Password).setValidators(validatorDefaultCredentials);
+    this.form.get(Controls.Password).updateValueAndValidity();
+
+    this.form.get(Controls.Project).setValidators(validatorDefaultCredentials);
+    this.form.get(Controls.Project).updateValueAndValidity();
+
+    this.form.get(Controls.ProjectID).setValidators(validatorDefaultCredentials);
+    this.form.get(Controls.ProjectID).updateValueAndValidity();
+
+    this.form.get(Controls.ApplicationCredentialID).setValidators(validatorAppCredentials);
+    this.form.get(Controls.ApplicationCredentialID).updateValueAndValidity();
+
+    this.form.get(Controls.ApplicationCredentialSecret).setValidators(validatorAppCredentials);
+    this.form.get(Controls.ApplicationCredentialSecret).updateValueAndValidity();
```

Comment:
There is definitely too much of the same here. It can be shortened with some array or map.

Reply:
Done